src/vivarium/framework/resource.py

```diff
@@ -0,0 +1,147 @@
+and in turn ask for ordered sequences of these resources according to their
+dependencies or raise exceptions if this is not possible.
+
+"""
+from types import MethodType
+from typing import List, Any, Iterable
+import warnings
+
+import networkx as nx
+
+from vivarium.exceptions import VivariumError
+
+
+class ResourceError(VivariumError):
+    """Error raised when a dependency requirement is violated"""
+    pass
+
+
+RESOURCE_TYPES = {'value', 'value_source', 'value_modifier', 'column', 'stream'}
+
+
+class ResourceProducer:
+
+    def __init__(self, resource_type: str, resource_names: List[str], producer: MethodType, dependencies: List[str]):
+        self.resource_type = resource_type
+        self.resource_names = resource_names
+        self.producer = producer
+        self.dependencies = dependencies
+
+    def __repr__(self):
+        resources = ','.join([f'{self.resource_type}.{name}' for name in self.resource_names])
+        return f'ResourceProducer({resources})'
+
+
+class EmptySet:
+
+    def add(self, item: Any):
+        pass
+
+    def __contains__(self, item: Any) -> bool:
+        return False
+
+
+class ResourceGroup:
+
+    def __init__(self, phase: str, single_producer: bool = False):
+        self.phase = phase
+        # One initializer per component, maybe multiple value producers
+        self.producer_components = set() if single_producer else EmptySet()
+        self.resources = {}
+        self._graph = None
+
+    @property
+    def graph(self):
+        if self._graph is None:
+            self._graph = self._to_graph()
+        return self._graph
+
+    def add_resources(self, resource_type: str, resource_names: List[str],
+                      producer: MethodType, dependencies: List[str]):
+        if resource_type not in RESOURCE_TYPES:
+            raise ResourceError(f'Unknown resource type {resource_type}.  Permitted types are {RESOURCE_TYPES}.')
+        if resource_type == 'column':
+            if producer.__self__.name in self.producer_components:
+                raise ResourceError  # Component has more than one producer for resource type ...
+            self.producer_components.add(producer.__self__.name)
+
+        producer = ResourceProducer(resource_type, resource_names, producer, dependencies)
+        for resource_name in resource_names:
+            key = f'{resource_type}.{resource_name}'
+            if key in self.resources:
+                raise ResourceError  # More than one producer for resource ...
+            self.resources[key] = producer
+
+    def __iter__(self) -> Iterable[MethodType]:
+        try:
+            sorted_nodes = nx.algorithms.topological_sort(self.graph)
+        except nx.NetworkXUnfeasible:
+            raise ResourceError(f'The resource group {self.phase} contains at least one cycle.')
+
+        return iter([r.producer for r in sorted_nodes if r.resource_type == 'column'])
+
+    def _to_graph(self) -> Iterable[ResourceProducer]:
+        g = nx.DiGraph()
```

Comment:
Restructured this to build the full dependency graph and have the iterator filter to column type things.  This makes much more compelling and informative dependency plots.